### ISSUE TITLE
Problem with return type assignment of Golang calling conventions

### ISSUE THREAD

**Issue by felberj** [NationalSecurityAgency/ghidra#692]
I am writing a plugin to assist reversing Golang binaries: https://github.com/felberj/gotools/

I have this problem where a function returns something, but the callee does not assign it to its local variables.

Example code: https://github.com/felberj/gotools/blob/master/testdata/functions.go
Compile it with: `GOOS=linux go build -gcflags '-N -l' functions.go`

Looking at the function main.A2r2:

```
func A2r2(a, b int) (int, int) {
	fmt.Println("A2R2", a, b)
	return 1, 2
}
```

The function gets its arguments as follows

| location | what | 
|---|---|
| Stack[0x8]:8 |   param1 | 
| Stack[0x10]:8 | param2 | 
| Stack[0x18]:8 | return1 | 
| Stack[0x20]:8 | return 2 | 

because I do not know how to return multiple values, I just return a struct:

  | /ret_5018 Unaligned Structure ret_5018 { undefined8 8 null "" undefined8 8 null "" } Size = 16 Actual Alignment = 1 |   | Stack[0x18]:16
-- | -- | -- | --
1 | undefined8 | param_1 | Stack[0x8]:8
2 | undefined8 | param_2 | Stack[0x10]:8

<img width="689" alt="image" src="https://user-images.githubusercontent.com/6639926/59566797-3feb6980-9065-11e9-8fca-662e74afafd3.png">

go__stdcall is defined here: https://github.com/felberj/gotools/blob/master/data/languages/golang1.12.cspec#L36

A2r2 looks ok when decompiled:

```c
ret_5018 main.A2r2(undefined8 param_1,undefined8 param_2)

{
  ulong *puVar1;
  ret_5018 rVar2;
  long in_FS_OFFSET;
  undefined8 uStack0000000000000018;
  undefined8 uStack0000000000000020;
  undefined *local_38;
  undefined1 *local_30;
  undefined *local_28;
  undefined8 local_18;
  undefined8 uStack16;
  
  puVar1 = (ulong *)(*(long *)(in_FS_OFFSET + 0xfffffff8) + 0x10);
  if (&local_18 < (undefined8 *)*puVar1 || &local_18 == (undefined8 *)*puVar1) {
    runtime.morestack_noctxt();
    rVar2 = main.A2r2(param_1,param_2);
    return rVar2;
  }
  // SNIP
  return (ret_5018)(ZEXT816(1) << 0x40);
}
```

But the callee (main.main) does not assign the result of it:

```c
void main.main(void)

{
  ulong *puVar1;
  long in_FS_OFFSET;
  long local_78;
  long local_70;
  long local_68;
  long lVar2;
  long local_60;
  long lVar3;
  
  puVar1 = (ulong *)(*(long *)(in_FS_OFFSET + 0xfffffff8) + 0x10);
  if (register0x00000020 < (undefined *)*puVar1 ||
      (undefined *)register0x00000020 == (undefined *)*puVar1) {
    runtime.morestack_noctxt();
    main.main();
    return;
  }
  main.A0r0();
  main.A1r0(0);
  local_78 = 0; // this is the first params of A4r0
  local_70 = 1; // this is the second params of A4r0
  local_68 = 2; // this is the third params of A4r0
  local_60 = 3; // this is the fourth params of A4r0
  main.A4r0(0,1,2,3);
  main.A0r4();  // no assignment
  lVar2 = local_68;
  lVar3 = local_60;
  main.A2r2(0,1); // no assignment
  if (local_78 + local_70 + local_68 + local_60 + lVar2 + lVar3 == 0) {
    os.Exit(1);
  }
  os.Exit(2);
  return;
}
```

Do you know how to solve this problem? I have the following assumptions but I do not know how to verify them:

- The return value + arguments are in the frame of the callee, not sure whether this is a problem
- Ghidra has a problem with the return on the stack, probably because my cspecs are wrong?

**Comment by c3rb3ru5d3d53c** [comment]
I can confirm this still does not work on 10.1.4

**Comment by dev747368** [comment]
In general, golang param handling in Ghidra is going to be a bit messy.

We've got an analyzer working its way through the pipeline before its ready for release.

A couple of things that helps the decompiler produce better code (inside the called function) is to double-book the stack storage areas of return values, and of params that are passed via registers (that get 'spilled' if needed to the caller-allocated stack location).

By 'double-book', I mean allocate a local variable that occupies the same stack locations (via custom storage) as the return value/spill location. 

You won't be able to create a ghidra call spec that truly matches go's real behavior.  Maybe for very simple functions it will match.

Also of note in the OP's example listing, its showing the nasty self-recursive call at the top of main.  I've seen that (in 1.18 binaries)  caused by a jmp being incorrectly marked as a conditional call by the shared return analyzer.

**Comment by dev747368** [comment]
There have been a few fixes to the way the decompiler handles storage for structs that have recently landed (most recent was 74e1bbdb57788c05d3598a4740327d5878d0fc63) that were necessary when working on our internal golang support.

You should see a much improved result when decompiling go func params and return values.

I'm going to mark this as closed as I'm pretty sure its been handled.
